### PR TITLE
Java projects shouldn't be able to launch with Firefox #248

### LINKED ISSUE #248: [Firefox Debug] Java projects shouldn't be able to launch with Firefox
With a java project opened, right-clicking on the editor or project and selecting `Debug as...` gives Firefox as a launch option. This doesn't make sense for java projects and must be removed.  
image

### PR DESCRIPTION
**Pull request by AObuchow**
fix  #248

Signed-off-by: Andrew Obuchowicz <aobuchow@redhat.com>

### CODE REVIEW

**Comment by AObuchow** [comment]
Something broke during the rebase. No longer able to launch firefox debugger. Looking into it now.

**Comment by AObuchow** [comment]
> Something broke during the rebase. No longer able to launch firefox debugger. Looking into it now.

Nevermind, was using the wrong target platform. The tests are failing however.

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 1]
```diff
@@ -545,11 +545,25 @@
...
                   </iterate>
                </with></enablement>
          </contextualLaunch>
       </shortcut>
+   </extension>
+    <extension
+         point="org.eclipse.core.runtime.adapters">
+      <factory
+            adaptableType="org.eclipse.core.resources.IResource"
+            class="org.eclipse.wildwebdeveloper.debug.firefox.FirefoxLaunchableAdapterFactory">
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: This is not really related to firefox, as we'll probably have the same logic for Chrome.
It could be something like "HTMLLaunchableAdapterFactory"

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package org.eclipse.wildwebdeveloper.debug.firefox;
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: Missing copyright header.

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 3]
```diff
@@ -545,11 +545,25 @@
                   <iterate>
                      <adapt
                            type="org.eclipse.core.resources.IResource">
+                        <test
+                              forcePluginActivation="true"
+                              property="org.eclipse.wildwebdeveloper.isHTMLProject">
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: At the moment, I don't think the launch shortcuts allow to properly handle projects, but work fine with files.
So I'd suggest that instead of checking for HTML project, we currently only check for HTML files. We can add projects in a separate commit,

File: org.eclipse.wildwebdeveloper/plugin.xml [hunk 4]
```diff
@@ -560,6 +574,13 @@
             properties="isNodeResource"
             type="java.lang.Object">
       </propertyTester>
+      <propertyTester
+            class="org.eclipse.wildwebdeveloper.debug.firefox.Tester"
+            id="org.eclipse.wildwebdeveloper.propertyTester.isHTMLProject"
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/plugin.xml]: Should be more something like `isHTMLLaunchable` so we can reuse it for Chrome support later.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import org.eclipse.core.runtime.IAdapterFactory;
+import org.eclipse.debug.ui.actions.ILaunchable;
+
+public class FirefoxLaunchableAdapterFactory implements IAdapterFactory {
+	
+	private static final ILaunchable DUMMY = new ILaunchable() {
+	};
+
+	@Override
+	public <T> T getAdapter(Object adaptableObject, Class<T> adapterType) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java]: Usually, first check is for the adpterType

```java
   if (!ILaunchable.class.equals(adapterType)) {
      return null;
   }
```

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+		IResource resource = Adapters.adapt(adaptableObject, IResource.class);
+		if (adapterType.equals(ILaunchable.class) && Tester.isHTMLProject(resource.getProject())) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java]: We also need to return some ILaunchable for HTML files.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.eclipse.core.expressions.PropertyTester;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IAdaptable;
+
+public class Tester extends PropertyTester {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: Please give is a more explicit name, like `IsLaunchableHTMLTester`

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	}
+
+	public static boolean isHTMLProject(IProject p) {
+		if (p == null || !p.isAccessible()) {
+			return false;
+		}
+		try {
+			for (IResource projItem : p.members()) {
+				if (projItem.getName().equals("index.html")) { //$NON-NLS-1$
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: The launch shortcut doesn't (yet) contain logic about detecting the `index.html`. Is this logic included in the language servers themselves? Ie, if provided a folder, do they automatically start the `index.html` ? That would actually be great and may automatically get things working better in Wild Web Developer.

**Comment by mickaelistria** [comment]
Also, can you please make it 2 commits: one about adding the ILaunchable where is makes sense (so the Debug As context-menu shows in Wild Web Developer), and another one about not showing firefox on non related resources?

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+		if (o instanceof IResource) {
+			return (IResource) o;
+		} else if (o instanceof IAdaptable) {
+			return ((IAdaptable) o).getAdapter(IResource.class);
+		} else {
+			return null;
+		}
+	}
+
+	public static boolean isHTMLProject(IProject p) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: Maybe this can simply be `new FirefoxRunDebugLaunchShortcut().canLaunch()` to improve factorization.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java (hunk 8 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: I don't think the debug adapter (I assume this is what you mean by language server) has any logic to automatically start the `index.html`. In my testing, trying to open a folder in Firefox presents the user with a directory explorer.

The logic in `Tester.java` is only used to determine if it makes sense for a file or project to be debuged in Firefox. 

The actual launch shortcut will most likely need logic to find the `index.html` and set that as the `file` parameter in `FirefoxRunDABDebugDelegate`.

File: org.eclipse.wildwebdeveloper/plugin.xml (hunk 3 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/plugin.xml]: I agree. I'll make it check if the user is trying to launch a HTML file rather than checking if the file belongs to a project that contains a `index.html`.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java (hunk 8 above)
Comment (mickaelistria) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: > The actual launch shortcut will most likely need logic to find the index.html and set that as the file parameter in FirefoxRunDABDebugDelegate.

Exactly, and I propose we add this logic later to focus on getting the shortcut shown only where it works at the moment (`.html` files).

Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: Agreed. Working on it now :)

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java (hunk 6 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java]: So I should check `IsLaunchableHTMLTester.isHTMLFile(resource)` and return the ILaunchable?

Comment (mickaelistria) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxLaunchableAdapterFactory.java]: Sounds good. Let's try and see ;)

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java (hunk 9 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/Tester.java]: Forgot about this change, doing this now.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	
+	public static boolean isHTMLFile(IResource resource) {
+		if (resource == null || !resource.isAccessible()) {
+			return false;
+		}
+		return resource.getName().matches((".*\\.html$"));
+	}
+
+	public static boolean isHTMLProject(IProject p) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java]: As discussed earlier, I think we should get rid of it at the moment since we do not support that case correctly.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDebugLaunchShortcut.java [hunk 11]
```diff
@@ -31,6 +32,15 @@ public void configureLaunchConfiguration(File file, ILaunchConfigurationWorkingC
 		wc.setAttribute(FirefoxRunDABDebugDelegate.FILE, file.getAbsolutePath());
 		wc.setAttribute(FirefoxRunDABDebugDelegate.WORKING_DIRECTORY, file.getParentFile().getAbsolutePath());
 	}
+	
+	public static boolean canLaunch(IResource resource) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDebugLaunchShortcut.java]: This method is already defined in the parent class, no need to reimplement it.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+public class IsLaunchableHTMLTester extends PropertyTester {
+	private static final String PROPERTY_NAME = "isHTMLLaunchable"; //$NON-NLS-1$
+	@Override
+	public boolean test(Object receiver, String property, Object[] args, Object expectedValue) {
+		if (property.equals(PROPERTY_NAME)) {
+			IResource resource = Adapters.adapt(receiver, IResource.class);
+			if (resource == null) {
+				return false;
+			}
+			return FirefoxRunDebugLaunchShortcut.canLaunch(resource);
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java]: `new FirefoxRunDebugLaunchShortcut().canLaunch(resource.getLocation().toFile())` would allow to reuse already existing code.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDebugLaunchShortcut.java (hunk 11 above)
Comment (AObuchow) [reply org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/firefox/FirefoxRunDebugLaunchShortcut.java]: Ah my bad, I just realized. Fixing it now.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/HTMLLaunchableAdapterFactory.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	private static ILaunchable DUMMY = new ILaunchable() {
+	};
+
+	@Override
+	public <T> T getAdapter(Object adaptableObject, Class<T> adapterType) {
+		if (!ILaunchable.class.equals(adapterType)) {
+			return null;
+		} else {
+			IResource resource = Adapters.adapt(adaptableObject, IResource.class);
+			if (adapterType.equals(ILaunchable.class) && (new FirefoxRunDebugLaunchShortcut()).canLaunch(resource.getLocation().toFile())) {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/HTMLLaunchableAdapterFactory.java]: You don't need to repeat the condition on adapterType here.

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/HTMLLaunchableAdapterFactory.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+public class HTMLLaunchableAdapterFactory implements IAdapterFactory {
+
+	private static ILaunchable DUMMY = new ILaunchable() {
+	};
+
+	@Override
+	public <T> T getAdapter(Object adaptableObject, Class<T> adapterType) {
+		if (!ILaunchable.class.equals(adapterType)) {
+			return null;
+		} else {
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/HTMLLaunchableAdapterFactory.java]: The `else` here can be removed (`else`s after a `return` are useless).

File: org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+public class IsLaunchableHTMLTester extends PropertyTester {
+	private static final String PROPERTY_NAME = "isHTMLLaunchable"; //$NON-NLS-1$
+	@Override
+	public boolean test(Object receiver, String property, Object[] args, Object expectedValue) {
+		if (property.equals(PROPERTY_NAME)) {
+			IResource resource = Adapters.adapt(receiver, IResource.class);
+			if (resource == null) {
+				return false;
+			}
+			return (new FirefoxRunDebugLaunchShortcut()).canLaunch(resource.getLocation().toFile());
```
Comment (mickaelistria) [inline org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/debug/IsLaunchableHTMLTester.java]: parenthesis here are not necessay.

**Comment by mickaelistria** [comment]
Thanks!
